cl/phase1/execution_client/block_collector/persistent_block_collector.go

```diff
@@ -0,0 +1,366 @@
+
+	// Process blocks in batches (blocks are already sorted by key due to MDBX ordering)
+	blocksBatch := []*types.Block{}
+	inserted := uint64(0)
+
+	for _, b := range blocks {
+		block, err := p.decodeBlock(b.value)
+		if err != nil {
+			p.logger.Warn("[BlockCollector] Failed to decode block", "err", err)
+			continue
+		}
+		if block == nil {
+			continue
+		}
+
+		blocksBatch = append(blocksBatch, block)
+
+		if len(blocksBatch) >= batchSize {
+			if err := p.insertBatch(ctx, blocksBatch, &inserted); err != nil {
+				return err
+			}
+			blocksBatch = []*types.Block{}
+		}
+	}
+
+	// Insert remaining blocks
+	if len(blocksBatch) > 0 {
+		if err := p.insertBatch(ctx, blocksBatch, &inserted); err != nil {
+			return err
+		}
+	}
+
+	// Close the database and remove the directory
+	p.db.Close()
+	p.db = nil
+
+	if err := dir.RemoveAll(p.persistDir); err != nil {
+		p.logger.Warn("[BlockCollector] Failed to remove database directory", "err", err)
+	}
+
+	p.logger.Info("[BlockCollector] Flush complete", "blocksInserted", inserted)
+
+	return nil
+}
+
+func (p *PersistentBlockCollector) decodeBlock(v []byte) (*types.Block, error) {
+	if len(v) == 0 {
+		return nil, nil
+	}
+
+	v, err := utils.DecompressSnappy(v, false)
+	if err != nil {
+		return nil, err
+	}
+
+	version := clparams.StateVersion(v[0])
+	parentRoot := common.BytesToHash(v[1:33])
+	requestsHash := common.Hash{}
+
+	if version >= clparams.ElectraVersion {
+		requestsHash = common.BytesToHash(v[33:65])
+		v = v[65:]
+	} else {
+		v = v[33:]
+	}
+
+	executionPayload := cltypes.NewEth1Block(version, p.beaconChainCfg)
+	if err := executionPayload.DecodeSSZ(v, int(version)); err != nil {
+		return nil, err
+	}
+
+	body := executionPayload.Body()
+	txs, err := types.DecodeTransactions(body.Transactions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Skip genesis block
+	if executionPayload.BlockNumber == 0 {
+		return nil, nil
+	}
+
+	header, err := executionPayload.RlpHeader(&parentRoot, requestsHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return types.NewBlockFromStorage(executionPayload.BlockHash, header, txs, nil, body.Withdrawals, nil), nil
+}
+
+func (p *PersistentBlockCollector) insertBatch(ctx context.Context, blocksBatch []*types.Block, inserted *uint64) error {
+	p.logger.Info("[BlockCollector] Inserting blocks",
+		"from", blocksBatch[0].NumberU64(),
+		"to", blocksBatch[len(blocksBatch)-1].NumberU64())
+
+	if err := p.engine.InsertBlocks(ctx, blocksBatch, true); err != nil {
+		p.logger.Warn("[BlockCollector] Failed to insert blocks", "err", err)
+		return err
+	}
+
+	*inserted += uint64(len(blocksBatch))
+	p.logger.Info("[BlockCollector] Inserted blocks", "progress", blocksBatch[len(blocksBatch)-1].NumberU64())
+
+	lastBlockHash := blocksBatch[len(blocksBatch)-1].Hash()
+	currentHeader, err := p.engine.CurrentHeader(ctx)
+	if err != nil {
+		p.logger.Warn("[BlockCollector] Failed to get current header", "err", err)
```

Comment:
on `err` you will have `isForkchoiceNeeded := currentHeader == nil` which seems unexpected?